File: scripts/mark_check.py
#!/usr/bin/env python3
"""Drives the reader's own marks through the real application, end to end.

Usage:
    scripts/mark_check.py <app-binary|tpdf.exe> <file.pdf> [--timeout SECONDS]

**What this covers that nothing else did.** `viewer_check.py` drives a real window
and builds its own `Viewer` with no edit model behind it, so a drag draws a
preview and commits nothing; the unit tests import modules and never `App.svelte`.
Between those two is the wiring that turns a command into a mark --- an object
literal binding the viewer's callbacks to the functions that reach the model ---
and on 2026-08-22 a reader found a defect living in exactly that gap: a shape
drawn on the last page of a document was dropped with no command sent and no
message shown, while all sixteen gates stayed green.

So this launches the app with a document, drives the route a reader takes, and
asserts against **the model** --- the marks that came back over the IPC boundary,
not the viewer that produced the gesture. `src/lib/markcheck.ts` holds the checks
and the argument for each; what this script adds is the launch and the reading of
the transcript, neither of which can be arranged from inside the process.

Takes the **binary**, not a `.app` bundle: unlike the open check there is no
Launch Services route here, and the document is handed over in `argv` like any
other spike entry point. On macOS the binary inside a bundle still needs the
bundle identity for WKWebView, so pass
`…/bundle/macos/tpdf.app/Contents/MacOS/tpdf` rather than `target/release/tpdf`.

Requires an unlocked, unoccluded screen for the same reason the viewer check does:
a suspended WebKit page does not run the check slowly, it does not run it at all.
"""

import argparse
import os
import sys
from pathlib import Path

import harness_launch
from harness_launch import outcome_of, run_app
from live_output import stream_results
from stray import clear_strays
from webview_guard import require_visible_session

#: The check whose absence means the run never got a document, by name.
#:
#: Duplicated from `markcheck.ts`, and it is a coupling rather than an assertion
#: --- a rename there would leave this unreachable and silently restore the
#: transcript of eight failures it exists to explain. So its absence is reported
#: as a failure of *this script*, never treated as "the fixture is fine".
PRECONDITION = "a document is open to put a mark on"

#: The one check that would have caught the defect this harness was written for.
#:
#: Named here so a run that somehow stopped printing it fails rather than passing
#: with one fewer line. A check that silently disappears is the failure mode this
#: repository records under "a check named by its position in a list".
KEYSTONE = "and it is recorded on the page it was pressed on"


def launch(binary: str, pdf: str, timeout: float) -> tuple[int, str]:
    """Runs the check once, returning its exit code and transcript.

    The launch and the transcript reader are `harness_launch.py`'s; what is left
    here is the one environment variable this check needs.
    """
    env = dict(os.environ, TPDF_MARKCHECK="1")
    done = run_app([binary, pdf], env=env, timeout=timeout)
    return done.code, done.out


def named_checks(out: str) -> "str | None":
    """The two checks this harness will not call a run green without.

    Read after the summary and the exit code have been compared and before the
    count of failures is reported, which is what `harness_launch.report`'s
    `extra` seam is for --- reporting "3 of 7 checks failed" here would point at
    the code when the fixture is what is wrong.

    The precondition is read rather than assumed. Without a document every check
    below it skips, and a transcript of skips with a green summary is exactly
    what a fixture problem looks like -- so it is named as one instead of being
    reported as a pass.

    And the keystone is the one assertion the harness exists for. A skip there is
    a legitimate outcome of the code path -- no mark was made, so there is no
    page to compare -- and it is not a legitimate outcome of a *run*: it means
    the thing this was built to check did not get checked.
    """
    got = outcome_of(out, PRECONDITION)
    if got != "OK":
        return f"the run never opened a document ({PRECONDITION!r}: {got})"
    keystone = outcome_of(out, KEYSTONE)
    if keystone != "OK":
        return f"the page-identity check did not run green ({keystone})"
    return None


def report(code: int, out: str) -> bool:
    """Prints the transcript, and says whether it is readable and green.

    The three grounds a run is refused on whatever it printed --- no summary
    line, a summary disagreeing with the exit code, a summary that is not all
    green --- are `harness_launch.report`'s, and the argument for each is written
    out there. This adds the two checks read by name, and the passing count on
    the way out: one run, one line, where a multi-phase script prints its own
    summary instead.
    """
    return harness_launch.report(out, code, extra=named_checks, announce=True)


#: A transcript of a clean run, as `Report` prints one. The self-test's control.
#:
#: Hand-written rather than captured, and that is the point: it is what the
#: script *claims* to accept, so a change to the reader that broke the shape has
#: something to fail against. The names match `markcheck.ts`; the padding does
#: not, deliberately, because `outcome_of` must not depend on a column width.
GREEN = """[OK]   a document is open to put a mark on           8 page(s)
[OK]   the add-comment command runs and arms the pointer   armed: note
[OK]   a comment placed by a press reaches the model    mark 1, a note
[OK]   and it is recorded on the page it was pressed on slot 0 is page 1; the mark says 1
[OK]   a shape drawn on the last page reaches the model mark 2 on page 8
7/7 checks passed
"""


def self_test() -> int:
    """Every way this script must refuse a transcript, and the one it accepts.

    **A reader that accepts everything is the instrument this repository has been
    caught by most.** `report` has four independent grounds for refusal and each
    one exists because the reassuring branch is the wrong answer: a run with no
    summary looks exactly like a silent success, `AppHandle::exit` does not set
    an exit code so the two numbers can disagree, a transcript of skips with a
    green summary is a fixture problem rather than a pass, and the keystone
    check skipping means the thing this harness was built for did not run.

    None of that needs a screen, which is the other reason it is here: the launch
    half cannot be exercised on a locked machine and this half can be exercised
    anywhere, so the script is never entirely unproved.
    """
    import io
    import contextlib

    def verdict(code: int, out: str) -> bool:
        with contextlib.redirect_stdout(io.StringIO()):
            return report(code, out)

    cases = [
        ("a clean transcript is accepted", True, 0, GREEN),
        (
            "a run with no summary line is refused, not read as silence",
            False,
            0,
            GREEN.replace("7/7 checks passed\n", ""),
        ),
        (
            "a summary that disagrees with the exit code is refused",
            False,
            1,
            GREEN,
        ),
        (
            "a failing summary is refused",
            False,
            1,
            GREEN.replace("7/7 checks passed", "6/7 checks passed"),
        ),
        (
            "a run that never opened a document is refused",
            False,
            0,
            GREEN.replace("[OK]   a document is open", "[FAIL] a document is open"),
        ),
        (
            "a skipped page-identity check is refused, not counted as green",
            False,
            0,
            GREEN.replace(
                "[OK]   and it is recorded on the page",
                "[SKIP] and it is recorded on the page",
            ),
        ),
    ]

    ok = True
    for name, want, code, out in cases:
        got = verdict(code, out)
        if got == want:
            print(f"[OK]   {name}")
        else:
            print(f"[FAIL] {name}: expected {want}, got {got}")
            ok = False

    # And that the name lookup does not depend on the column a name happens to
    # land in --- the failure `outcome_of`'s own comment describes, which reads
    # as good news because a name it cannot find is not a name that failed.
    widened = GREEN.replace(
        "[OK]   and it is recorded on the page it was pressed on ",
        "[OK]   and it is recorded on the page it was pressed on, in the model's own numbering    ",
    )
    if outcome_of(widened, KEYSTONE) == "OK":
        print("[OK]   a name found by prefix, whatever it is padded to")
    else:
        print("[FAIL] the reader is keyed on a column width")
        ok = False

    return 0 if ok else 2


def main() -> int:
    # Before anything prints: a redirected run is block-buffered otherwise, and
    # a partial transcript is then an empty file. See `live_output`.
    stream_results()
    if "--self-test" in sys.argv[1:]:
        # Before the argument parser, which requires a binary and a document ---
        # neither of which this needs, and demanding them would make the one part
        # runnable without a screen the part that needs a build.
        return self_test()
    parser = argparse.ArgumentParser()
    parser.add_argument("binary")
    parser.add_argument("pdf")
    parser.add_argument("--timeout", type=float, default=180.0)
    args = parser.parse_args()

    if not require_visible_session():
        return 1

    # A stray from an earlier run silently absorbs every later launch on Windows,
    # where single-instance forwards the argv and exits. See `stray`.
    clear_strays(Path(args.binary))

    pdf = str(Path(args.pdf).resolve())
    code, out = launch(args.binary, pdf, args.timeout)
    return 0 if report(code, out) else 1


if __name__ == "__main__":
    sys.exit(main())

File: src/lib/markcheck.ts
/**
 * A functional check of the reader's own marks, inside the real application.
 *
 * ## Why this exists, and it is not thoroughness
 *
 * On 2026-08-22 a reader reported that a shape drawn on a page vanished when
 * they let go of the button. It did. `Viewer.onDrawn` reported the page's *id*
 * and `Edits.mark` read that number as a *slot*, so a mark was written to the
 * next page along --- and on the last page, where there is no next slot, to
 * nowhere at all, with no command sent and no refusal shown.
 *
 * **Every gate was green, at four layers.** `edits.test.ts` asserts that the
 * model call sends an id, and it did. `viewerdraw.test.ts` asserts that the
 * viewer reports an id, and it did. `viewer_check.py` drives a real window ---
 * and builds its own `Viewer` with no model behind it, so a drag draws its
 * preview and commits nothing. The defect was in none of those halves. It was
 * in the *join*, which is an object literal in `App.svelte`: a file no unit test
 * imports, no window harness constructs, and nothing had ever executed with a
 * document open.
 *
 * So this check exists to run the route a reader takes and nothing else:
 *
 * > command → gesture on the real viewer → callback → edit model → overlay
 *
 * It asserts against **the model**, never against the viewer that produced the
 * gesture. Reading the answer back out of the thing under test is the
 * writer-agreeing-with-its-own-reader failure this repository has been caught by
 * twice; here the two ends are genuinely independent, because the model is in
 * Rust and heard about the mark over an IPC boundary.
 *
 * ## The one assertion that would have caught it
 *
 * *"and it is recorded on the page it was pressed on"* derives the expected page
 * **id** from the model's own page list and the viewer's own slot, and compares
 * it against the id the model filed the mark under. Under the shipped defect
 * those differ by one on every document, which is why a fixture of any length
 * catches it --- and on the last page there is no mark at all, which the check
 * above it says.
 *
 * ## What it deliberately does not do
 *
 * It does not check what a mark *looks like*: `viewer_check.py`'s overlay phase
 * measures ink per kind against a mark handed straight to `setMarks`, and
 * `annot-probe` measures the saved file's pixels. One ink reading is here, and
 * its job is only to say that a mark the model has reaches the screen --- the
 * last hop of the chain, which a model assertion cannot see.
 */

import { invoke } from "@tauri-apps/api/core";

import { pause, Report, settle as settleFor } from "./checkreport";
import type { MarkView, PageView } from "./pages";
import type { Viewer } from "./viewer";

/** How long any single wait may take before the check gives up. */
const TIMEOUT_MS = 30_000;

/**
 * How long to wait before deciding that *no* mark was made.
 *
 * A negative assertion cannot settle on a condition --- there is nothing to wait
 * for --- so it waits out the round trip instead. Generous, because what is
 * being ruled out is a command reaching Rust and coming back, and a check that
 * declared "nothing happened" before the reply landed would be a control that
 * passes whatever the code does.
 */
const QUIET_MS = 1_500;

const report = new Report();

const settle = (predicate: () => boolean) => settleFor(predicate, TIMEOUT_MS);

const check = (name: string, ok: boolean, detail: string) => report.check(name, ok, detail);

/** What the check needs from the application, in the application's own terms. */
export interface MarkCheckHost {
  /**
   * Runs a command through the registry, exactly as the palette does.
   *
   * **Not the underlying action.** What is being tested is the chain a reader
   * sets off, and the registry is its first link --- a check that called
   * `armDraw` directly would skip the enablement guard, the command's own
   * argument handling and the wiring in between, which is three quarters of
   * what can be wrong.
   */
  run: (id: string) => boolean;
  /** The live viewer, for where things are on screen and for the overlay. */
  viewer: () => Viewer | null;
  /**
   * The element the viewer was built on, which is where a press has to land.
   *
   * Handed over rather than read off the viewer, exactly as `sessioncheck.ts`
   * takes it: the viewer does not expose its own root, and adding an accessor so
   * that a harness can reach around to the element the application already holds
   * would widen the public surface for no one else's benefit.
   */
  root: () => HTMLElement | null;
  /** The marks the **model** holds. The independent end of every assertion. */
  marks: () => readonly MarkView[];
  /** The pages the model holds, so an expected page id can be derived. */
  pages: () => readonly PageView[];
  /** The path that is open, so the check can say it had a document. */
  path: () => string;
}

/**
 * Runs the check if `TPDF_MARKCHECK` is set, then exits the process.
 *
 * Returns `false` when it was not requested, so the caller carries on into the
 * real application. Called at the end of the boot, after a document handed over
 * on the command line has opened --- the check needs one and does not open it
 * itself, for `sessioncheck.ts`'s reason: the open is part of the application
 * and a harness that replaced it would be testing something else.
 */
export async function runMarkCheckIfRequested(host: MarkCheckHost): Promise<boolean> {
  const mode = await invoke<string | null>("markcheck_mode");
  if (!mode) return false;

  try {
    await run(host);
  } catch (e) {
    report.check("the check ran", false, String(e));
  }

  await report.finish();
  return true;
}

/** A press and release at one point, with no movement between them. */
function click(root: HTMLElement, x: number, y: number): void {
  root.dispatchEvent(
    new PointerEvent("pointerdown", {
      button: 0,
      pointerId: 1,
      clientX: x,
      clientY: y,
      bubbles: true,
    }),
  );
  root.dispatchEvent(
    new PointerEvent("pointerup", { pointerId: 1, clientX: x, clientY: y, bubbles: true }),
  );
}

/** A press, a move and a release --- the gesture every shape tool reads. */
function drag(
  root: HTMLElement,
  from: { x: number; y: number },
  to: { x: number; y: number },
): void {
  root.dispatchEvent(
    new PointerEvent("pointerdown", {
      button: 0,
      pointerId: 1,
      clientX: from.x,
      clientY: from.y,
      bubbles: true,
    }),
  );
  root.dispatchEvent(
    new PointerEvent("pointermove", {
      pointerId: 1,
      clientX: to.x,
      clientY: to.y,
      bubbles: true,
    }),
  );
  root.dispatchEvent(
    new PointerEvent("pointerup", {
      pointerId: 1,
      clientX: to.x,
      clientY: to.y,
      bubbles: true,
    }),
  );
}

/** The ids in the model's page list, in reading order. */
const idsOf = (pages: readonly PageView[]) => pages.map((page) => page.id);

/** The mark the model gained since `before`, or `null` for none. */
function added(before: ReadonlySet<number>, now: readonly MarkView[]): MarkView | null {
  return now.find((mark) => !before.has(mark.id)) ?? null;
}

/**
 * The fraction of a rectangle on the overlay that carries any of our ink.
 *
 * The alpha channel rather than a colour, for `viewercheck.ts`'s reason: the
 * overlay is transparent where nothing was painted, and keying on a particular
 * colour would make this a check on the palette instead.
 */
function inked(viewer: Viewer, box: { left: number; top: number; right: number; bottom: number }): number | null {
  const canvas = viewer.overlaySurface;
  const ctx = canvas.getContext("2d", { willReadFrequently: true });
  if (!ctx) return null;
  const dpr = window.devicePixelRatio || 1;
  const x0 = Math.max(0, Math.round(box.left * dpr));
  const y0 = Math.max(0, Math.round(box.top * dpr));
  const x1 = Math.min(canvas.width, Math.round(box.right * dpr));
  const y1 = Math.min(canvas.height, Math.round(box.bottom * dpr));
  if (x1 - x0 < 2 || y1 - y0 < 2) return null;
  const { data } = ctx.getImageData(x0, y0, x1 - x0, y1 - y0);
  let hit = 0;
  for (let at = 3; at < data.length; at += 4) {
    if ((data[at] ?? 0) > 8) hit++;
  }
  return hit / (data.length / 4);
}

async function run(host: MarkCheckHost): Promise<void> {
  const opened = await settle(() => host.path() !== "" && host.viewer() !== null);
  check(
    "a document is open to put a mark on",
    opened,
    opened ? `${host.pages().length} page(s)` : "nothing opened",
  );
  if (!opened) return;

  const viewer = host.viewer();
  const root = host.root();
  if (!viewer || !root) {
    check("the viewer hands over a surface to press on", false, "no surface");
    return;
  }
  await settle(() => viewer.idle);

  // Mid-window, well inside whatever page is on screen. Nothing here depends on
  // the point being over text: a comment is dropped on the paper and a box is
  // dragged across it, which is exactly why these tools have no `hasSelection`
  // guard.
  const bounds = root.getBoundingClientRect();
  const mid = { x: bounds.left + bounds.width * 0.4, y: bounds.top + bounds.height * 0.4 };

  // --- The control, and it goes first ------------------------------------
  //
  // Every assertion below says "a mark appeared". Without this, all of them are
  // satisfied by an application that makes a mark on any press at all --- and
  // the press position is shared, so this is the same press.
  {
    const before = new Set(host.marks().map((mark) => mark.id));
    click(root, mid.x, mid.y);
    await pause(QUIET_MS);
    const stray = added(before, host.marks());
    check(
      "a press with nothing armed marks nothing",
      stray === null,
      stray ? `made a ${stray.kind}` : `${host.marks().length} mark(s), unchanged`,
    );
  }

  // --- A comment, placed by a press --------------------------------------
  {
    const armed = host.run("edit.addComment");
    check("the add-comment command runs and arms the pointer", armed && viewer.drawArmed === "note", `armed: ${viewer.drawArmed}`);

    const slot = viewer.position.page;
    const before = new Set(host.marks().map((mark) => mark.id));
    click(root, mid.x, mid.y);

    const landed = await settle(() => added(before, host.marks()) !== null);
    const mark = added(before, host.marks());
    check(
      "a comment placed by a press reaches the model",
      landed && mark?.kind === "note",
      mark ? `mark ${mark.id}, a ${mark.kind}` : "the model gained nothing",
    );

    if (mark) {
      // **The assertion the whole file is for.** The expected id comes from the
      // model's own page list at the viewer's own slot; the actual id is what
      // the model filed the mark under. Two ends that only agree if every hop
      // between them carried the same page --- and under the defect this check
      // was written for they differ by one on every document there is.
      const want = host.pages()[slot]?.id;
      check(
        "and it is recorded on the page it was pressed on",
        want !== undefined && mark.page === want,
        `slot ${slot} is page ${want}; the mark says ${mark.page} (ids: ${idsOf(host.pages()).slice(0, 6).join(", ")})`,
      );
    } else {
      report.skip(
        "and it is recorded on the page it was pressed on",
        "no mark was made, so there is no page to compare",
      );
    }
  }

  // --- A shape on the last page ------------------------------------------
  //
  // The case that produced no mark at all rather than one in the wrong place:
  // the id of the last page is one past the end of the slot list, so the lookup
  // that should not have been there answered `undefined` and the method returned
  // without sending anything. A one-page document exercises it too --- which is
  // most of what a reader opens, and is why the defect was reported so quickly.
  {
    const last = host.pages().length - 1;
    viewer.goToPage(last);
    await settle(() => viewer.idle);

    const armed = host.run("edit.drawBox");
    check("the draw-box command runs and arms the pointer", armed && viewer.drawArmed === "square", `armed: ${viewer.drawArmed}`);

    const slot = viewer.position.page;
    const before = new Set(host.marks().map((mark) => mark.id));
    drag(root, { x: mid.x, y: mid.y }, { x: mid.x + 120, y: mid.y + 90 });

    const landed = await settle(() => added(before, host.marks()) !== null);
    const mark = added(before, host.marks());
    check(
      "a shape drawn on the last page reaches the model",
      landed && mark?.kind === "square",
      mark
        ? `mark ${mark.id} on page ${mark.page}`
        : `nothing was made on slot ${slot} of ${host.pages().length}`,
    );

    if (mark) {
      const want = host.pages()[slot]?.id;
      check(
        "and that one names the last page too",
        want !== undefined && mark.page === want,
        `slot ${slot} is page ${want}; the mark says ${mark.page}`,
      );

      // The last hop, and the only one a model assertion cannot see: a mark the
      // model holds has to reach the screen. Read over the viewer's own anchor
      // for the mark, so a rotated or cropped page is sampled where the mark
      // actually is rather than where an independent derivation thinks it is.
      await settle(() => viewer.idle);
      const anchor = viewer.markAnchor(mark.id);
      const ink = anchor ? inked(viewer, anchor) : null;
      if (ink === null) {
        report.skip(
          "and the overlay draws it",
          anchor ? "the overlay has no readable 2d context" : "the mark has no anchor",
        );
      } else {
        check(
          "and the overlay draws it",
          ink > 0.01,
          `${(ink * 100).toFixed(1)}% of the mark's own rectangle is inked`,
        );
      }

      // --- Moving it, and undoing that ------------------------------------
      const home = [...mark.quads];
      const centre = anchor
        ? { x: (anchor.left + anchor.right) / 2, y: (anchor.top + anchor.bottom) / 2 }
        : mid;
      drag(root, centre, { x: centre.x - 60, y: centre.y - 40 });

      const moved = await settle(() => {
        const now = host.marks().find((item) => item.id === mark.id);
        return now !== undefined && now.quads[0] !== home[0];
      });
      const now = host.marks().find((item) => item.id === mark.id);
      check(
        "a mark dragged across the page moves in the model",
        moved && now !== undefined,
        now ? `left ${home[0]?.toFixed(1)} -> ${now.quads[0]?.toFixed(1)}` : "the mark is gone",
      );

      if (moved) {
        check(
          "and undo puts it back where it was",
          host.run("edit.undo") &&
            (await settle(() => {
              const back = host.marks().find((item) => item.id === mark.id);
              return back !== undefined && back.quads[0] === home[0];
            })),
          `back to ${host.marks().find((item) => item.id === mark.id)?.quads[0]?.toFixed(1)} from ${home[0]?.toFixed(1)}`,
        );
      } else {
        report.skip("and undo puts it back where it was", "nothing moved, so there is nothing to undo");
      }
    } else {
      for (const name of [
        "and that one names the last page too",
        "and the overlay draws it",
        "a mark dragged across the page moves in the model",
        "and undo puts it back where it was",
      ]) {
        report.skip(name, "no mark was made, so there is nothing to look at");
      }
    }
  }
}
